### PR TITLE
[Dy2Stat] Add debugging and logging mechanism for dygraph to static

### PR DESCRIPTION
**Pull request by liym27**
### PR types

 New features
### PR changes

Others
### Describe

Add debugging and logging mechanism for dygraph to static graph
1. Add logs in the conversion of function, and unified log information, clear and easy to understand.
 For example:
```
2020-08-26 11:09:42,088-Level 2: Whitelist: <paddle.fluid.dygraph.nn.Linear object at 0x13306dd70> is part of Paddle module and does not have to be transformed.
```

2. Add function `set_verbosity(level=0)` for user to set the verbosity level for dygraph to static graph.
    There are two means to set the logging verbosity:
     1. Call function `set_verbosity`
     2. Set environment variable `TRANSLATOR_VERBOSITY`
 
3. Add function `set_code_level(level)` for user to set the level to print code from specific Ast Transformer.

### CODE REVIEW

File: python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py [hunk 1]
```diff
@@ -19,21 +19,21 @@
...
 import gast
-
 from paddle.fluid.dygraph.dygraph_to_static.assert_transformer import AssertTransformer
 from paddle.fluid.dygraph.dygraph_to_static.basic_api_transformer import BasicApiTransformer
 from paddle.fluid.dygraph.dygraph_to_static.break_continue_transformer import BreakContinueTransformer
 from paddle.fluid.dygraph.dygraph_to_static.call_transformer import CallTransformer
 from paddle.fluid.dygraph.dygraph_to_static.cast_transformer import CastTransformer
 from paddle.fluid.dygraph.dygraph_to_static.ifelse_transformer import IfElseTransformer
 from paddle.fluid.dygraph.dygraph_to_static.list_transformer import ListTransformer
+from paddle.fluid.dygraph.dygraph_to_static.logging_utils import TranslatorLogger
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py]: Move the import line 29 between line 35 and line 36.

File: python/paddle/fluid/dygraph/dygraph_to_static/call_transformer.py [hunk 2]
```diff
@@ -62,6 +62,12 @@ def visit_Call(self, node):
             return node
 
         func_str = ast_to_source_code(node.func).strip()
+
+        # NOTE(liym27): Don't convert `pad.set_trace` even if the convertion doesn't work finally, because
+        # it is clearer to see where it is called from.
+        if "pdb.set_trace" in func_str:
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/call_transformer.py]: Should we make this string a constant?

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+LogicalTransformer = 6
+LoopTransformer = 7
+IfElseTransformer = 8
+AssertTransformer = 9
+PrintTransformer = 10
+CallTransformer = 11
+CastTransformer = 12
+AllTransformer = 12
+
+_transformer_name_to_level = {
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: Can we remove these code? If we develop new Transformer and forget to add these map/integer, we made error. I comment at line 117 "def log_transformed_code" to say the implementation I thought.

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        self.logger.error(msg, *args, **kwargs)
+
+    def warn(self, msg, *args, **kwargs):
+        self.logger.warn(msg, *args, **kwargs)
+
+    def log(self, level, msg, *args, **kwargs):
+        if self.has_verbosity(level):
+            self.logger.log(level, msg, *args, **kwargs)
+
+    def log_transformed_code(self, level, ast_node, *args, **kwargs):
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: 1. You level is not an integer but a string, level at other places is integer, you had better to rename it "transformer_name" or something similar to it.

2. As I commented, could we delete "_transformer_name_to_level", just keep an integer and you increment the integer by 1 every time you call this method? Then your "has_code_level" can compare the integer and the code level set by user.

3. msg can add "transformer_name". For example:

msg = "After the {} -st/rd/th {}, transformed code:\n{}".format(the_integer, transformer_name, ast_to_source_code(ast_node))

Then users or Paddle develop can know the mapping of their level to transformer_name, and transformer code

File: python/paddle/fluid/dygraph/dygraph_to_static/call_transformer.py (hunk 2 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/call_transformer.py]: done

File: python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py (hunk 1 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py]: done

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 5]
```diff
@@ -61,6 +58,24 @@ def is_paddle_func(func):
     return m is not None and m.__name__.startswith("paddle")
 
 
+def is_unsupported(func):
+    # TODO(liym27): A better way to do this.
+    if any(func in m.__dict__.values()
+           for m in (collections, pdb, copy, inspect, re, six, numpy)):
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Is it better to define a inner constant variable, such as `BUILTIN_LIKELY_MODULES` to deal this?

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 6]
```diff
@@ -95,13 +110,14 @@ def dyfunc(x):
           #  [1. 1. 1.]]
 
     """
+    translator_logger.log(1, "Convert call: convert {}.".format(func))
     func_self = None
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:118]: `call` is an AST but not clear concept for users.  Adding `ast.Call` or `Convert callable object:`

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 7]
```diff
@@ -61,6 +58,24 @@ def is_paddle_func(func):
     return m is not None and m.__name__.startswith("paddle")
 
 
+def is_unsupported(func):
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:62]: Not import. It seems `is_unsupported`  is not a good name that will confuse developers, such as what is unsupported?

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+VERBOSITY_ENV_NAME = 'TRANSLATOR_VERBOSITY'
+DEFAULT_VERBOSITY = 0
+
+
+def synchronized(func):
+    def wrapper(*args, **kwargs):
+        with threading.Lock():
+            return func(*args, **kwargs)
+
+    return wrapper
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py:35]: After this pr is merged, we can consider to reuse code in ProgramTranslator.

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+
+class TranslatorLogger(object):
+    """
+    class for Logging and debugging during the tranformation from dygraph to static graph.
+    The object of this class is a singleton.
+    """
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py:42]: We can add comment here to indicate it is a Singleton， such as `A singleton class`

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+    @transformed_code_level.setter
+    def transformed_code_level(self, level):
+        self.check_level(level, _transformer_name_to_level)
+        self._transformed_code_level = level
+
+    def check_level(self, level, name_to_level_dict):
+        if isinstance(level, (six.integer_types, type(None))):
+            rv = level
+        elif str(level) == level:
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: Why not use `isinstance(str, six.string_types)`?

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+            if level not in name_to_level_dict:
+                raise ValueError("Unknown level: %r" % level)
+            rv = name_to_level_dict[level]
+        else:
+            raise TypeError("Level not an integer or a valid string: %r" %
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: Missing `is`.

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        level(int): The verbosity level. The larger value idicates more verbosity.
+            The default value is 0, which means no logging.
+    Examples:
+        .. code-block:: python
+
+            import os
+            import paddle.fluid as fluid
+
+            fluid.dygraph.dygraph_to_static.set_verbosity(1)
+            # The verbosity level is now 1
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py:149]: Consider to expose interface as `paddle.jit.set_verbosity()`?

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        level(int): The level to print code.
+
+    Examples:
+        .. code-block:: python
+
+            import os
+            import paddle.fluid as fluid
+            from paddle.fluid.dygraph.dygraph_to_static import logging_utils
+
+            logging_utils.set_code_level(logging_utils.CastTransformer)
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: `paddle.jit.set_code_level`?

**Review by Aurelius84 — COMMENTED** [review]
Great job~~

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 3 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: done

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 4 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: Thanks, done

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 5 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Done

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 6 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:118]: Thanks. use `Convert callable object`:
`translator_logger.log(1, "Convert callable object: convert {}.".format(func))`

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 8 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py:35]: OK

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 9 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py:42]: Yes, `The object of this class is a singleton.` has indicated it is a singleton

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 10 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: Thanks, it has been removed

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 13 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: done

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 12 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py:149]: done

File: python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py (hunk 11 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/logging_utils.py]: done

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 7 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:62]: Add doc for this function "Checks whether the func is supported by dygraph to static graph."

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by Aurelius84 — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
LGTM~

**Review by lanxianghit — APPROVED** [review]

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**